torch/csrc/jit/passes/xnnpack_rewrite.cpp

```diff
@@ -0,0 +1,121 @@
+        %packed_weight_bias = xnnpack::linear_prepack(%weight, %bias)
+        %res = xnnpack::linear_packed(%input, %packed_weight_bias)
+        return (%res))";
+  std::string linear_pattern = R"(
+    graph(%input, %weight, %bias):
+        %r = aten::linear(%input, %weight, %bias)
+        return (%r))";
+  std::string xnnpack_pattern = R"(
+    graph(%input, %weight, %bias):
+        %packed_weight_bias = xnnpack::linear_prepack(%weight, %bias)
+        %res = xnnpack::linear_packed(%input, %packed_weight_bias)
+        return (%res))";
+  std::string linear_pattern_wo_bias = R"(
+    graph(%input, %weight):
+        %bias: Tensor? = prim::Constant()
+        %res = aten::linear(%input, %weight, %bias)
+        return (%res))";
+  std::string xnnpack_pattern_wo_bias = R"(
+    graph(%input, %weight):
+        %bias: Tensor? = prim::Constant()
+        %packed_weight_bias = xnnpack::linear_prepack(%weight, %bias)
+        %res = xnnpack::linear_packed(%input, %packed_weight_bias)
+        return (%res))";
+
+  auto filter = [](const Match& match,
+                   const std::unordered_map<std::string, Value*>& vmap) {
+    const auto& match_vmap = match.values_map;
+    auto linear_value = match_vmap.at(vmap.at("linear"));
+    auto func_name = getFuncName(linear_value);
+    if (func_name == "linear") {
+      return true;
+    }
+    return false;
+  };
+
+  SubgraphRewriter linear_call_fn_rewriter;
+  linear_call_fn_rewriter.RegisterRewritePattern(linear_before_inline, xnnpack_pattern_before_inline);
+  linear_call_fn_rewriter.runOnGraph(graph, filter);
+
+  SubgraphRewriter linear_rewriter;
+  linear_rewriter.RegisterRewritePattern(linear_pattern, xnnpack_pattern);
+  linear_rewriter.runOnGraph(graph);
+
+  SubgraphRewriter linear_rewriter_wo_bias;
+  linear_rewriter_wo_bias.RegisterRewritePattern(linear_pattern_wo_bias, xnnpack_pattern_wo_bias);
+  linear_rewriter_wo_bias.runOnGraph(graph);
+}
+
+void insertXNNPACKConv2dOp(std::shared_ptr<Graph>& graph) {
+  // Replace _convolution with conv2d
+  graph_rewrite_helper::replaceConvolutionWithConv2d(graph);
+
+  std::string conv_2d_pattern = R"(
+    graph(%input, %weight, %bias, %stride:int[], %padding:int[], %dilation:int[], %groups:int):
+        %r = aten::conv2d(%input, %weight, %bias, %stride, %padding, %dilation, %groups)
+        return (%r) )";
+
+  std::string xnnpack_conv2d_pattern = R"(
+    graph(%input, %weight, %bias, %stride:int[], %padding:int[], %dilation:int[], %groups:int):
+        %packed_weight_bias = xnnpack::conv2d_prepack(%weight, %bias, %stride, %padding, %dilation, %groups)
+        %r = xnnpack::conv2d_packed(%input, %packed_weight_bias)
+        return (%r) )";
+
+  SubgraphRewriter rewriter;
+  rewriter.RegisterRewritePattern(conv_2d_pattern, xnnpack_conv2d_pattern);
+  rewriter.runOnGraph(graph);
+}
+
+} // namespace
+
+
+void insertXNNPACKOps(std::shared_ptr<Graph>& graph) {
+  ConstantPooling(graph);
+  ConstantPropagation(graph);
```

Comment:
also, we plan to remove these from our code, it's better not to depend on them, could you check if removing them also works?

Reply:
Do you mean constant propagation? To replace aten::convolution with conv2d you need this. Else it does not work.

Reply:
yeah, we plan to make our patterns not rely on constant prop.

Do you mean fuse linear? replace aten::_convolution should work without constant prop

Reply:
I remember it did not work, because you need to check the parameters values. Some values pass through prim::ListConstruct and without const prop. you do not see the values.
For example output_padding value. This is the one it was not able to find without const prop.